### PR TITLE
Adding Custom Rules to Device Propagation

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67575
* #67500
* __->__ #66973
* #66948

Differential Revision: [D32497549](https://our.internmc.facebook.com/intern/diff/D32497549)

### CODE REVIEW

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175958).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175958).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175958).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32497549).

File: torch/csrc/jit/passes/dtype_analysis.cpp [hunk 1]
```diff
@@ -240,8 +290,24 @@ struct DtypePropagationPass {
...
+      return rule(n);
+    }
     return tryApplyDtypeMetaTensor(n);
   }
+
+  void buildDtypeRuleRegistry() {
+    // building a registry for all of the custom dtype rules
+    dtype_prop_registry_ = std::make_unique<OperatorMap<DtypePropRule>>();
+
+    dtype_prop_registry_->insert(
```
Comment (eellison) [inline torch/csrc/jit/passes/dtype_analysis.cpp:305]: Could we add tests for the new operators that we're using here ? these are still untested

**Pull request closed without merging**